### ISSUE TITLE
Generic class decorator does not work with forward-referencing?

### ISSUE THREAD

**Issue by indigoviolet** [microsoft/pyright#3290]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
Forward referencing isn't working in a generic class decorator:

```python
from __future__ import annotations

from typing import TypeVar, Generic, Type

_T = TypeVar("_T")


class Deco(Generic[_T]):
    # I've removed the code for which this was made generic, to simplify the example
    def __call__(self, cls: Type[_T]) -> Type[_T]:
        return cls


@Deco[Foo]() <---   /tmp/test_pyright.py:13:7 - error: "Foo" is not defined (reportUndefinedVariable)
class Foo:
    pass
```

**To Reproduce**
See above

**Expected behavior**
I expected no errors since I'm using the forward-referencing mechanic from PEP563.

[mypy](https://mypy-play.net/?mypy=latest&python=3.10&gist=f197697b980f520c3aa80fe611b868a1) seems not to have this error.

**Screenshots or Code**
See above

**VS Code extension or command-line**

```
❯ ~/.emacs.d/.local/etc/lsp/npm/pyright/bin/pyright --version                                          
pyright 1.1.231
```
command-line, or emacs via lsp-pyright.

**Additional context**

**Comment by erictraut** [comment]
Pyright is correct here.

The forward-referencing mechanism from PEP 563 applies only to type annotations, not to expressions that are part of the normal code that is necessarily executed by the Python interpreter. The code above will not run without generating a runtime exception (`NameError: name 'Foo' is not defined`). Mypy doesn't detect these types of errors but pyright does.

**Comment by indigoviolet** [comment]
Yuck, I should have checked that. Thanks for the amazingly quick response and this amazing tool.
